### PR TITLE
Refactor WatchService

### PR DESCRIPTION
**Pull request by J-N-K**
Depends on #2994 

This replaces the `AbstractWatchService` with a new `WatchService` interface and implementation that can be consumed as OSGi service. It makes use of a library that does a lot of what we currently do (hashing, removing duplicate events) and also implements a native watch service on MacOS which decreases response time to file changes.

- [x] Implement WatchService
- [x] Adjust implementations
- [x] Remove Java17 changes to core (should be merged in a separate PR).
- [x] Resolve itests (not done yet as it will create unnecessary merge conflicts)
- [x] Adjust usages in openhab-addons (https://github.com/openhab/openhab-addons/pull/14004)

@wborn Since we discussed this issue in the past, WDYT about the design?

### CODE REVIEW

File: bom/compile/pom.xml
Comment (jlaur) [inline bom/compile/pom.xml:325]: When resuming PR, you probably want latest version, so just a reminder - currently:
```suggestion
      <version>0.17.1</version>
```

Comment (J-N-K) [reply bom/compile/pom.xml:345]: Needs an upgrade in openhab-osgiify before

Comment (jlaur) [reply bom/compile/pom.xml:345]: openhab/openhab-osgiify#37 was just merged. 👍

**Comment by florian-h05** [comment]
@J-N-K What is the state here? I mean how much finished is this one?

**Comment by J-N-K** [comment]
Code-wise complete, but I'm pretty sure there are some minor tweaks needed.

**Comment by wborn** [comment]
One of the minor tweaks is to fix this test:

```
TEST org.openhab.core.model.core.internal.folder.FolderObserverTest#testModification() <<< ERROR: 
Expected: is <true>
     but: was <false>
java.lang.AssertionError: 
Expected: is <true>
     but: was <false>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
	at org.openhab.core.model.core.internal.folder.FolderObserverTest.lambda$7(FolderObserverTest.java:219)
	at org.openhab.core.test.java.JavaTest.waitForAssert(JavaTest.java:245)
	at org.openhab.core.test.java.JavaTest.waitForAssert(JavaTest.java:213)
	at org.openhab.core.test.java.JavaTest.waitForAssert(JavaTest.java:177)
	at org.openhab.core.model.core.internal.folder.FolderObserverTest.testModification(FolderObserverTest.java:219)
```

**Comment by J-N-K** [comment]
This needs changes after #3255, so I would like to see that one merged first (it is easier to to the changes here). I'll also have another look at the tests then.

**Comment by J-N-K** [comment]
> One of the minor tweaks is to fix this test:
> 
> ```
> TEST org.openhab.core.model.core.internal.folder.FolderObserverTest#testModification()
> ```

I refactored it to a unit test. We want to test the `FolderObserver`, not the `WatchService`. It's clearly a timing issue, as it sometimes occurs, sometimes not and never on my local builds.

**Comment by J-N-K** [comment]
@wborn This seems to be ready now.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/internal/loader/DefaultScriptFileWatcher.java
Comment (wborn) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/internal/loader/DefaultScriptFileWatcher.java:52]: Why remove `@Override` if it still overrides the parent method?

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java
Comment (wborn) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:70]: If it is no longer an abstract class the name `AbstractScriptFileWatcher` is wrong.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptDependencyTracker.java
Comment (wborn) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptDependencyTracker.java:68]: `Kind` is an enum so I think you can compare using `==` instead.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java
Comment (wborn) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:232]: `Kind` is an enum so I think you can compare using `==` instead.

**Review by wborn — CHANGES_REQUESTED** [review]
It looks good in general! :+1:
I do already have a few minor review comments. :slightly_smiling_face:

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/internal/loader/DefaultScriptFileWatcher.java
Comment (J-N-K) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/internal/loader/DefaultScriptFileWatcher.java]: I have removed the methods, the `activate`/`deactivate` methods are called automatically.

**Comment by florian-h05** [comment]
@wborn 
FYI, I think your review was addressed.

**Comment by florian-h05** [comment]
@openhab/core-maintainers Would be nice if this gets merged soon. Thanks!

**Review by wborn — APPROVED** [review]
Many thanks!

**Pull request merged**
